include/boost/unordered/detail/foa.hpp

```diff
@@ -1023,6 +1024,47 @@ inline void prefetch(const void* p)
 #endif    
 }
 
+template<typename Allocator>
+struct is_std_allocator:std::false_type{};
+
+template<typename T>
+struct is_std_allocator<std::allocator<T>>:std::true_type{};
+
+/* std::allocator::construct marked as deprecated */
+#if defined(_LIBCPP_SUPPRESS_DEPRECATED_PUSH)
+_LIBCPP_SUPPRESS_DEPRECATED_PUSH
+#elif defined(_STL_DISABLE_DEPRECATED_WARNING)
+_STL_DISABLE_DEPRECATED_WARNING
+#elif defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable:4996)
+#endif
+
+template<typename Allocator,typename Ptr,typename... Args>
+struct alloc_has_construct
```

Comment:
Maybe the better thing to do is we instead enhance `boost::allocator_construct_n()` to call `memcpy()` when `alloc_has_construct` is `false` and `T` is trivially copyable.

Reply:
@joaquintides Personally, I'd be interested in seeing this. How do you feel? I don't mind holding back the PR to wait for an implementation.

Reply:
No need to hold back the PR. :) Can always refactor later.

Reply:
I'm not sure this will help here, because the bucket array can have "holes" in it, unconstructed elements, but is still copied with a single memcpy. (Not 100% positive though.)

Reply:
Maybe just expose `allocator_has_construct` as a supported, user-facing, primitive.

Reply:
It's as Peter says, the `std::memcpy` call goes through an array with uninitialized portions. Having `allocator_has_construct` exposed would save us some some code duplication.